_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+---
+layout: post
+title:  "The HTTP part 6 - HTTP Client"
+date:   2018-11-01 13:10:31 +0530
+categories: ["programming", "http", "server", "kotlin"]
+comments: true
+---
+Previously we’ve built multithreaded server with a connection pooling. We’ve tested it using Apache HTTP Client. Since we’ve learned how pooling works, it’s a good time to build our own HTTP client with the connection pooling!
+
+## The requirements
+We need to meet the following requirements:
+
+* #### Connections are pooled
+
+
+We will reuse a connection for multiple requests. The time a connection is kept open without any activity is configured by the _keep alive timeout_.
+
+* #### Maximum connections per destination
+
+
+
+There should be a limit of opened connection per destination, which is a pair of host and port. To explain this, let’s say we set it to 2. We send a request that creates the first connection. If we send another one before first is completed then the new connection is created.
+If 2 requests are pending and we send yet another one, then we have to wait for one of them to finish. We wait for maximum time configured by the _connection request timeout_.
+
+Now, let’s get back to the situation without any opened connection. We send a request that opens a connection. We send another one but before that, the previous request is completed so the connection is released back to the pool. This second request will lease and use this connection, therefore only 1 connection would be opened.
```

Comment:
Then we send another one, but before that the previous...

Reply:
(Quick disclaimer: I'm not a native english speaker so my comments about english language are just suggestions)

I would say that correct form here is: "without any opened connection**s**". Here is a stackexchange post about plural vs singular after "without any": https://ell.stackexchange.com/questions/124317/do-you-have-to-use-singular-and-plural-nouns-after-any

I would say that http servers are meant to handle multiple connections at the same time so this comment applies here:
`It would depend on what is expected. If only one error message may come up, but it is uncertain as to what the message would be, 'any message' is appropriate. If a number of undefined error messages could come up, 'any messages' is more appropriate.`

Reply:
I would go for something like `reuse previous connection`, not `lease`. If it's more common to talk about leasing a connection in that context then I'm probably wrong.

Reply:
Yeah, "reuse" is better in this context.
Thanks for the link to SE, it sounds better with "s" too :)